### PR TITLE
[Feature]  Support submit spark jar and sql tasks to yarn

### PR DESCRIPTION
**Pull request by ChengJie1053**
[WIP][Feature]  Support submit spark jar tasks to yarn

issus: https://github.com/apache/incubator-streampark/issues/3569

Submit the spark yarn jar task by referring to the flink module

maven compile command
mvn  -Pshaded,spark,webapp,dist -DskipTests clean install -e

image

image

image

Submit the spark sql task to yarn
image

CREATE TABLE jsonSourceTable (name VARCHAR(64))
USING org.apache.spark.sql.json
OPTIONS (
  path "hdfs:///test/cjtest/test.json"
);
CREATE TABLE jsonSinkTable (name VARCHAR(64))
USING org.apache.spark.sql.json
OPTIONS (
  path "hdfs:///test/cjtest/test2.json"
);
INSERT INTO jsonSinkTable SELECT * FROM jsonSourceTable

image

image

image

### CODE REVIEW

**Comment by wolfboys** [comment]
This is a very great feature! In 2024, we plan to support Spark. We had a discussion about Spark before:
https://github.com/apache/incubator-streampark/discussions/3043
 so we need to revisit that. Additionally, the project's CI failed. Can you address this issue first? 😊

**Comment by caicancai** [comment]
@ChengJie1053 Thank you for your contribution. This is a big new feature. We should send issues and emails to discuss it. At least let us know your plan.

**Comment by ChengJie1053** [comment]
> This is a very great feature! In 2024, we plan to support Spark. We had a discussion about Spark before: #3043 so we need to revisit that. Additionally, the project's CI failed. Can you address this issue first? 😊

Ok, thanks for reviewing the code

**Comment by ChengJie1053** [comment]
> Thank you for your contribution. This is a big new feature. We should send issues and emails to discuss it. At least let us know your plan.

Ok, thanks for reviewing the code

**Comment by wolfboys** [comment]
There are some conflicts, can you resolve them?

File: streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql [hunk 1]
```diff
@@ -556,4 +556,97 @@ create table if not exists `t_resource` (
...
+create table `t_spark_env` (
+                               `id` bigint not null auto_increment comment 'id',
+                               `spark_name` varchar(128) collate utf8mb4_general_ci not null comment 'spark instance name',
+                               `spark_home` varchar(255) collate utf8mb4_general_ci not null comment 'spark home path',
+                               `version` varchar(64) collate utf8mb4_general_ci not null comment 'spark version',
+                               `scala_version` varchar(64) collate utf8mb4_general_ci not null comment 'scala version of spark',
+                               `spark_conf` text collate utf8mb4_general_ci not null comment 'spark-conf',
+                               `is_default` tinyint not null default 0 comment 'whether default version or not',
+                               `description` varchar(255) collate utf8mb4_general_ci default null comment 'description',
+                               `create_time` datetime not null default current_timestamp comment 'create time',
```
Comment (wolfboys) [inline streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql]: `default current_timestamp` has syntax compatibility issue, change to : `default null`

File: streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql [hunk 2]
```diff
@@ -556,4 +556,97 @@ create table if not exists `t_resource` (
...
+                               `k8s_rest_exposed_type` tinyint default null,
+                               `jm_memory` int default null,
+                               `tm_memory` int default null,
+                               `total_task` int default null,
+                               `total_tm` int default null,
+                               `total_slot` int default null,
+                               `available_slot` int default null,
+                               `option_state` tinyint default null,
+                               `tracking` tinyint default null,
+                               `create_time` datetime not null default current_timestamp comment 'create time',
```
Comment (wolfboys) [inline streampark-console/streampark-console-service/src/main/assembly/script/schema/mysql-schema.sql]: default current_timestamp -> default null

**Comment by ChengJie1053** [comment]
Ok, I'll sort it out as soon as possible



---- Replied Message ----
| From | ***@***.***> |
| Date | 03/25/2024 10:40 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [apache/incubator-streampark] [WIP][Feature] Support submit spark jar and sql tasks to yarn (PR #3537) |

There are some conflicts, can you resolve them?

**Comment by ChengJie1053** [comment]
> There are some conflicts, can you resolve them?

Ok, I'll sort it out as soon as possible

**Comment by wolfboys** [comment]
Is the current code compatible with the Spark 2.x? Spark has Scala versions 2.11, 2.12, and 2.13, so should streamprak support all of these Scala versions? How to implement? It is an issue that needs to be discussed. We can continue to address these issues in the future.

**Comment by ChengJie1053** [comment]
> Is the current code compatible with the Spark 2.x? Spark has Scala versions 2.11, 2.12, and 2.13, so should streamprak support all of these Scala versions? How to implement? It is an issue that needs to be discussed. We can continue to address these issues in the future.

Currently only support spark 3.1.2 version, do not support multiple versions, you can refer to the flink module later, support spark multiple versions, but do not intend to support in the current pr, because the current pr code has changed a lot, the review code is not very convenient

**Comment by wolfboys** [comment]
> > Is the current code compatible with the Spark 2.x? Spark has Scala versions 2.11, 2.12, and 2.13, so should streamprak support all of these Scala versions? How to implement? It is an issue that needs to be discussed. We can continue to address these issues in the future.
> 
> Currently only support spark 3.1.2 version, do not support multiple versions, you can refer to the flink module later, support spark multiple versions, but do not intend to support in the current pr, because the current pr code has changed a lot, the review code is not very convenient

Ok,Thanks for your hard work, StreamPark has taken a new step in supporting apache spark

File: streampark-spark/streampark-spark-core/src/main/scala/org/apache/streampark/spark/core/util/SqlCommandParser.scala [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+  @Deprecated
+  case object BEGIN_STATEMENT_SET
+    extends SqlCommand("begin statement set", "BEGIN\\s+STATEMENT\\s+SET", Converters.NO_OPERANDS)
+
+  /** This is SQL Client's syntax, don't use in our platform. */
+  @Deprecated
+  case object END_STATEMENT_SET
+    extends SqlCommand("end statement set", "END", Converters.NO_OPERANDS)
+
+  // Since: 2.1.2 for flink 1.18
```
Comment (wolfboys) [inline streampark-spark/streampark-spark-core/src/main/scala/org/apache/streampark/spark/core/util/SqlCommandParser.scala]: flink ?

File: streampark-spark/streampark-spark-core/src/main/scala/org/apache/streampark/spark/core/util/SqlCommandParser.scala [hunk 4]
```diff
@@ -0,0 +1,658 @@
...
+import java.lang.{Boolean => JavaBool}
+import java.util.Scanner
+import java.util.regex.{Matcher, Pattern}
+
+import scala.annotation.tailrec
+import scala.collection.{immutable, mutable}
+import scala.collection.mutable.ListBuffer
+import scala.util.control.Breaks.{break, breakable}
+
+object SqlCommandParser extends Logger {
```
Comment (wolfboys) [inline streampark-spark/streampark-spark-core/src/main/scala/org/apache/streampark/spark/core/util/SqlCommandParser.scala:35]: Copy from flink Implementation and never used it?

Comment (ChengJie1053) [reply streampark-spark/streampark-spark-core/src/main/scala/org/apache/streampark/spark/core/util/SqlCommandParser.scala:35]: image

image


The SqlCommandParser is used here

File: streampark-spark/streampark-spark-core/src/main/scala/org/apache/streampark/spark/core/util/SqlCommandParser.scala [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+   * This is an equivalent feature with StatementSet in Table API.
+   * The STATEMENT SET syntax encloses one or more INSERT INTO statements.
+   * All statements in a STATEMENT SET block are holistically optimized and executed as a single Spark job.
+   * Joint optimization and execution allows for reusing common intermediate results and can therefore significantly
+   * improve the efficiency of executing multiple queries.
+   * </pre>
+   */
+  /** This is SQL Client's syntax, don't use in our platform. */
+  @Deprecated
+  case object BEGIN_STATEMENT_SET
```
Comment (wolfboys) [inline streampark-spark/streampark-spark-core/src/main/scala/org/apache/streampark/spark/core/util/SqlCommandParser.scala]: I see that many syntax is in FlinkSQL, is SparkSQL the same syntax?

Comment (ChengJie1053) [reply streampark-spark/streampark-spark-core/src/main/scala/org/apache/streampark/spark/core/util/SqlCommandParser.scala]: The syntax supported by FlinkSQL and SparkSQL should be different, I will modify the code

**Review by wolfboys — APPROVED** [review]
Overall looks good, thanks for your hard work

**Pull request merged**